### PR TITLE
Чернова Наталья. Задача 1. Вариант 24. Подсчет числа слов в строке

### PR DESCRIPTION
**Pull request by ChernovaNatalya**
**Описание последовательной задачи.**
Входные данные - строка, в которой нужно найти количество входящих слов.
Выходные данные - целочисленное значение - количество слов в строке.

**Описание параллельной задачи.**
Разделяем строку на несколько подстрок и поручаем эту задачу нескольким потокам. Разделение происходит равномерно. Затем каждый поток считает число слов в своей части строки. В итоге всё суммируется с помощью reduce.

### CODE REVIEW

File: tasks/mpi/chernova_n_word_count/src/ops_mpi.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  internal_order_test();
+  if (world.rank() == 0) {
+    return taskData->inputs_count[0] > 0 && taskData->outputs_count[0] == 1;
+  }
+  return true;
+}
+
+bool chernova_n_word_count_mpi::TestMPITaskParallel::run() {
+  internal_order_test();
+  std::cout << world.size() << std::endl;
```
Comment (VasenkovAA) [inline tasks/mpi/chernova_n_word_count/src/ops_mpi.cpp:99]: and you're sure these are the ones you want in the final version?

**Review by VasenkovAA — APPROVED** [review]

File: tasks/mpi/chernova_n_word_count/src/ops_mpi.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    input_[i] = tmp_ptr[i];
+  }
+  input_ = clean_string(input_);
+  return true;
+}
+
+bool chernova_n_word_count_mpi::TestMPITaskSequential::validation() {
+  internal_order_test();
+  return taskData->inputs_count[0] >= 0 && taskData->outputs_count[0] == 1;
+}
```
Comment (VlJulia) [inline tasks/mpi/chernova_n_word_count/src/ops_mpi.cpp:55]: Why ">=0"? Also what if the string is empty?

**Review by VlJulia — CHANGES_REQUESTED** [review]

**Review by VlJulia — APPROVED** [review]

File: tasks/mpi/chernova_n_word_count/src/ops_mpi.cpp (hunk 1 above)
Comment (ChernovaNatalya) [reply tasks/mpi/chernova_n_word_count/src/ops_mpi.cpp]: fixed it

File: tasks/mpi/chernova_n_word_count/src/ops_mpi.cpp (hunk 2 above)
Comment (ChernovaNatalya) [reply tasks/mpi/chernova_n_word_count/src/ops_mpi.cpp]: You're right. I fixed it and also added a test to check for empty strings.

**Review by allnes — APPROVED** [review]

File: tasks/mpi/chernova_n_word_count/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+
+  std::shared_ptr<ppc::core::TaskData> taskDataParallel = std::make_shared<ppc::core::TaskData>();
+
+  if (world.rank() == 0) {
+    taskDataParallel->inputs.emplace_back(reinterpret_cast<uint8_t *>(const_cast<char *>(in.data())));
+    taskDataParallel->inputs_count.emplace_back(in.size());
+    taskDataParallel->outputs.emplace_back(reinterpret_cast<uint8_t *>(out.data()));
+    taskDataParallel->outputs_count.emplace_back(out.size());
+    chernova_n_word_count_mpi::TestMPITaskParallel testTaskParallel(taskDataParallel);
+    ASSERT_FALSE(testTaskParallel.validation());
```
Comment (allnes) [inline tasks/mpi/chernova_n_word_count/func_tests/main.cpp:22]: I think empty string has 0 words

**Review by allnes — CHANGES_REQUESTED** [review]

File: tasks/mpi/chernova_n_word_count/func_tests/main.cpp (hunk 3 above)
Comment (ChernovaNatalya) [reply tasks/mpi/chernova_n_word_count/func_tests/main.cpp]: fixed it

File: tasks/mpi/chernova_n_word_count/include/ops_mpi.hpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+
+namespace chernova_n_word_count_mpi {
+
+std::vector<char> clean_string(const std::vector<char>& input);
+
```
Comment (aobolensk) [inline tasks/mpi/chernova_n_word_count/include/ops_mpi.hpp:18]: Please move test functions to the tests module

File: tasks/seq/chernova_n_word_count/func_tests/main.cpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  ASSERT_TRUE(testTaskSequential.validation());
+  testTaskSequential.pre_processing();
+  testTaskSequential.run();
+  testTaskSequential.post_processing();
+  ASSERT_EQ(out[0], 20);
+}
+
+TEST(Sequential_chernova_n_word_count, Test_five_words_with_space_in_the_end) {
+  std::vector<char> in;
+  std::string testString = "This is a test phrase           ";
```
Comment (aobolensk) [inline tasks/seq/chernova_n_word_count/func_tests/main.cpp:136]: Please, add random tests

File: tasks/mpi/chernova_n_word_count/func_tests/main.cpp [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    ASSERT_EQ(out[0], referenceWordCount[0]);
+  }
+}
+
+TEST(chernova_n_word_count_mpi, Test_five_words) {
+  boost::mpi::communicator world;
+  std::vector<char> in;
+  std::string testString = "This is a test phrase";
+  for (unsigned long int j = 0; j < testString.length(); j++) {
+    in.push_back(testString[j]);
```
Comment (aobolensk) [inline tasks/mpi/chernova_n_word_count/func_tests/main.cpp:51]: You can use std::copy or analogs here

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/mpi/chernova_n_word_count/include/ops_mpi.hpp (hunk 4 above)
Comment (ChernovaNatalya) [reply tasks/mpi/chernova_n_word_count/include/ops_mpi.hpp:18]: Sorry, but I can't move this function to the test module. I use this function to process the string in the algorithm itself. I have test functions that I use only in test module, but they are declared only there (function generateWords in perf_tests).

File: tasks/seq/chernova_n_word_count/func_tests/main.cpp (hunk 5 above)
Comment (ChernovaNatalya) [reply tasks/seq/chernova_n_word_count/func_tests/main.cpp:149]: Fixed it

File: tasks/mpi/chernova_n_word_count/src/ops_mpi.cpp [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+bool chernova_n_word_count_mpi::TestMPITaskSequential::pre_processing() {
+  internal_order_test();
+  input_ = std::vector<char>(taskData->inputs_count[0]);
+  spaceCount = 0;
+  auto* tmp_ptr = reinterpret_cast<char*>(taskData->inputs[0]);
+  for (unsigned i = 0; i < taskData->inputs_count[0]; i++) {
+    input_[i] = tmp_ptr[i];
+  }
+  input_ = clean_string(input_);
```
Comment (kmichaelk) [inline tasks/mpi/chernova_n_word_count/src/ops_mpi.cpp:46]: Do you process the whole string by single worker before actual task run? This time is comparable to the time it takes to count the number of words in a line in a single process, and taking into account the `erase` calls and reallocation overhead, it even exceeds it in the worst case - it defeats the purpose of parallelizing the task.

**Review by kmichaelk — CHANGES_REQUESTED** [review]

**Review by allnes — APPROVED** [review]

File: tasks/mpi/chernova_n_word_count/include/ops_mpi.hpp (hunk 4 above)
Comment (aobolensk) [reply tasks/mpi/chernova_n_word_count/include/ops_mpi.hpp:18]: Are you sure that this is correct?

**Review by aobolensk — APPROVED** [review]
